Fix and enhance xinfo command

jonaskaempf opened the pull request:
Fixes #434. Implement more stable ELF parsing using pyelftools.

Also print containing ELF sections.

Examples:
```
pwndbg> vmmap
LEGEND: STACK | HEAP | CODE | DATA | RWX | RODATA
          0x400000           0x41e000 r-xp    1e000 0      /bin/ls
          0x61d000           0x61e000 r--p     1000 1d000  /bin/ls
          0x61e000           0x61f000 rw-p     1000 1e000  /bin/ls
          0x61f000           0x620000 rw-p     1000 0      [heap]
    0x7ffff715a000     0x7ffff7172000 r-xp    18000 0      /lib/x86_64-linux-gnu/libpthread-2.23.so
    0x7ffff7172000     0x7ffff7371000 ---p   1ff000 18000  /lib/x86_64-linux-gnu/libpthread-2.23.so
    0x7ffff7371000     0x7ffff7372000 r--p     1000 17000  /lib/x86_64-linux-gnu/libpthread-2.23.so
    0x7ffff7372000     0x7ffff7373000 rw-p     1000 18000  /lib/x86_64-linux-gnu/libpthread-2.23.so
    0x7ffff7373000     0x7ffff7377000 rw-p     4000 0      
    0x7ffff7377000     0x7ffff737a000 r-xp     3000 0      /lib/x86_64-linux-gnu/libdl-2.23.so
    0x7ffff737a000     0x7ffff7579000 ---p   1ff000 3000   /lib/x86_64-linux-gnu/libdl-2.23.so
    0x7ffff7579000     0x7ffff757a000 r--p     1000 2000   /lib/x86_64-linux-gnu/libdl-2.23.so
    0x7ffff757a000     0x7ffff757b000 rw-p     1000 3000   /lib/x86_64-linux-gnu/libdl-2.23.so
    0x7ffff757b000     0x7ffff75e9000 r-xp    6e000 0      /lib/x86_64-linux-gnu/libpcre.so.3.13.2
    0x7ffff75e9000     0x7ffff77e9000 ---p   200000 6e000  /lib/x86_64-linux-gnu/libpcre.so.3.13.2
    0x7ffff77e9000     0x7ffff77ea000 r--p     1000 6e000  /lib/x86_64-linux-gnu/libpcre.so.3.13.2
    0x7ffff77ea000     0x7ffff77eb000 rw-p     1000 6f000  /lib/x86_64-linux-gnu/libpcre.so.3.13.2
    0x7ffff77eb000     0x7ffff79ab000 r-xp   1c0000 0      /lib/x86_64-linux-gnu/libc-2.23.so
    0x7ffff79ab000     0x7ffff7bab000 ---p   200000 1c0000 /lib/x86_64-linux-gnu/libc-2.23.so
    0x7ffff7bab000     0x7ffff7baf000 r--p     4000 1c0000 /lib/x86_64-linux-gnu/libc-2.23.so
    0x7ffff7baf000     0x7ffff7bb1000 rw-p     2000 1c4000 /lib/x86_64-linux-gnu/libc-2.23.so
    0x7ffff7bb1000     0x7ffff7bb5000 rw-p     4000 0      
    0x7ffff7bb5000     0x7ffff7bd4000 r-xp    1f000 0      /lib/x86_64-linux-gnu/libselinux.so.1
    0x7ffff7bd4000     0x7ffff7dd3000 ---p   1ff000 1f000  /lib/x86_64-linux-gnu/libselinux.so.1
    0x7ffff7dd3000     0x7ffff7dd4000 r--p     1000 1e000  /lib/x86_64-linux-gnu/libselinux.so.1
    0x7ffff7dd4000     0x7ffff7dd5000 rw-p     1000 1f000  /lib/x86_64-linux-gnu/libselinux.so.1
    0x7ffff7dd5000     0x7ffff7dd7000 rw-p     2000 0      
    0x7ffff7dd7000     0x7ffff7dfd000 r-xp    26000 0      /lib/x86_64-linux-gnu/ld-2.23.so
    0x7ffff7fcf000     0x7ffff7fd4000 rw-p     5000 0      
    0x7ffff7ff7000     0x7ffff7ffa000 r--p     3000 0      [vvar]
    0x7ffff7ffa000     0x7ffff7ffc000 r-xp     2000 0      [vdso]
    0x7ffff7ffc000     0x7ffff7ffd000 r--p     1000 25000  /lib/x86_64-linux-gnu/ld-2.23.so
    0x7ffff7ffd000     0x7ffff7ffe000 rw-p     1000 26000  /lib/x86_64-linux-gnu/ld-2.23.so
    0x7ffff7ffe000     0x7ffff7fff000 rw-p     1000 0      
    0x7ffffffde000     0x7ffffffff000 rw-p    21000 0      [stack]
0xffffffffff600000 0xffffffffff601000 r-xp     1000 0      [vsyscall]
```
```
pwndbg> xinfo 0x61e610
Extended information for virtual address 0x61e610:

  Containing mapping:
          0x61e000           0x61f000 rw-p     1000 1e000  /bin/ls

  Offset information:
         Mapped Area 0x61e610 = 0x61e000 + 0x610
       File (Memory) 0x61e610 = 0x400000 + 0x21e610
         File (Disk) 0x61e610 = [not file-backed]

 Containing ELF sections:
                .bss 0x61e610 = 0x61e600 + 0x10
pwndbg> xinfo 0x40cafe
Extended information for virtual address 0x40cafe:

  Containing mapping:
          0x400000           0x41e000 r-xp    1e000 0      /bin/ls

  Offset information:
         Mapped Area 0x40cafe = 0x400000 + 0xcafe
       File (Memory) 0x40cafe = 0x400000 + 0xcafe
         File (Disk) 0x40cafe = /bin/ls  + 0xcafe

 Containing ELF sections:
               .text 0x40cafe = 0x402a00 + 0xa0fe
```
```
pwndbg> xinfo 0x7ffff7fcf001
Extended information for virtual address 0x7ffff7fcf001:

  Containing mapping:
    0x7ffff7fcf000     0x7ffff7fd4000 rw-p     5000 0      

  Offset information:
         Mapped Area 0x7ffff7fcf001 = 0x7ffff7fcf000 + 0x1
pwndbg> xinfo 0x7ffff77eb440
Extended information for virtual address 0x7ffff77eb440:

  Containing mapping:
    0x7ffff77eb000     0x7ffff79ab000 r-xp   1c0000 0      /lib/x86_64-linux-gnu/libc-2.23.so

  Offset information:
         Mapped Area 0x7ffff77eb440 = 0x7ffff77eb000 + 0x440
       File (Memory) 0x7ffff77eb440 = 0x7ffff77eb000 + 0x440
         File (Disk) 0x7ffff77eb440 = /lib/x86_64-linux-gnu/libc-2.23.so + 0x440

 Containing ELF sections:
           .gnu.hash 0x7ffff77eb440 = 0x7ffff77eb2b8 + 0x188
pwndbg> xinfo 0x7ffff7ffd020
Extended information for virtual address 0x7ffff7ffd020:

  Containing mapping:
    0x7ffff7ffd000     0x7ffff7ffe000 rw-p     1000 26000  /lib/x86_64-linux-gnu/ld-2.23.so

  Offset information:
         Mapped Area 0x7ffff7ffd020 = 0x7ffff7ffd000 + 0x20
       File (Memory) 0x7ffff7ffd020 = 0x7ffff7dd7000 + 0x226020
         File (Disk) 0x7ffff7ffd020 = /lib/x86_64-linux-gnu/ld-2.23.so + 0x26020

 Containing ELF sections:
            .got.plt 0x7ffff7ffd020 = 0x7ffff7ffd000 + 0x20
```

Commit:
Instead of unstable parsing of readelf output, use the elftools ELF wrapper for parsing PT_LOAD segments

Commit:
Fix #434 xinfo command doesn't show File(Disk) info on non-PIE binaries

Also remove some trailing whitespace

Also fix another bug in xinfo; now it can show the disk offset of all
mmap files, not just the primary executable

Commit:
New xinfo feature: Print containing ELF sections for file-backed addresses

Commit:
Only print header for ELF sections if at least 1 section contains the address

Commit:
Fix bug in section offset calculation when printing containing ELF sections

disconnect3d commented:
Hey,

Thanks for the PR!

Via [pyelftools github repo]:
> It works with Python versions 2.7 and 3.x (x >= 2)

This means the pwndbg will probably not work on Python 3.0 and 3.1. I have a bit mixed feelings about that but I don't have Py 3.0-3.1 anywhere so it is okay for me. 

@zachriggle What do you think?

---

There is one small issue I found when testing it - output formatting:
```
pwndbg> xinfo
Extended information for virtual address 0x4048c5:

  Containing mapping:
          0x400000           0x41c000 r-xp    1c000 0      /bin/ls

  Offset information:
         Mapped Area 0x4048c5 = 0x400000 + 0x48c5
       File (Memory) 0x4048c5 = 0x400000 + 0x48c5
         File (Disk) 0x4048c5 = /bin/ls   + 0x48c5

 Containing ELF sections:
               .text 0x4048c5 = 0x4028a0 + 0x2025
```
(the `+` after `/bin/ls` could be at the same position as above)

disconnect3d commented on pwndbg/elf.py:
```diff
@@ -68,6 +70,59 @@ def read(typ, address, blob=None):
     return obj
 
 
+@pwndbg.proc.OnlyWhenRunning
```
Hmm the function seems it doesn't use anything that requires the program to be running, maybe we can drop that decorator from here?

jonaskaempf replied:
Sure, I have no opinion on that decorator. Just kept it because it was in the original code, I believe

disconnect3d commented on pwndbg/elf.py:
```diff
@@ -68,6 +70,59 @@ def read(typ, address, blob=None):
+    segments = []
+    with open(local_path, 'rb') as f:
+        elffile = ELFFile(f)
+        is_pic = elffile.header['e_type'] == 'ET_DYN'
```
Maybe we should expose `pwndbg.elf.is_pic`?

jonaskaempf replied:
It could make sense to exposed a cached view of ELF properties. However, I think it could be a bit ambiguous maybe to put it as a global property on pwndbg.elf. I realise that usually you want to know whether the main binary is PIC or not (since libraries will today always be PIC). However, one of the bugs in the original xinfo fell into this trap seeing as it always opened the main executable in readelf when finding ELF segments. However, xinfo can of course be applied to something other than the main executable, so that sometimes produced an exception.

Pushing the logic into a separate memoized function in pwndbg.elf could make a lot of sense though. I can craft another commit implementing this

disconnect3d commented on pwndbg/elf.py:
```diff
@@ -68,6 +70,59 @@ def read(typ, address, blob=None):
+
+
+@pwndbg.proc.OnlyWhenRunning
+def get_containing_sections(elf_filename, vaddr, loadaddr):
```
We can probably make both functions to be `memoize.reset_on_objfile` or at least make another function that would get `ELFFile` and return a list of sections? or sections that are interesting for us (not to call `ELFFile` and open file multiple times)

jonaskaempf replied:
Agreed. Can craft another commit for that

jonaskaempf commented:
In terms of the pyelftools compatibility: this package was already a pwndbg dependency, so I guess it does not change pwndbg compliancy. Also, I've never seen python3.{0,1} used in stable environments

Commit:
Refactor ELF file parsing helpers for cleaner separtion of ELF metadata parsing and enrichment, and a specific use scenario (getting a list of segments/sections containing a given virtual addr). Also makes implementing caching parse results easier

Adjust xinfo command to these API changes

jonaskaempf commented:
Added a fairly substantial re-write of the ELF parsing helpers and appropriate adjustments of the xinfo command implementation.

Also fixed a few bugs in xinfo formatting.

Added another (potential) line of output: Before, we did not print a separate line specifying the offset of the vaddr to a given ELF segment address. A given address may fall in the "slack" space between the end of a PT_LOAD segment and a page boundary. It is now possible to see such cases in the absences of a "File (Segment)" line. Maybe we want to add output in the negative case similar to the "File (Disk) ... = [not file-backed]" instance?

Updated examples:

```
pwndbg> vmmap
LEGEND: STACK | HEAP | CODE | DATA | RWX | RODATA
          0x400000           0x41e000 r-xp    1e000 0      /bin/ls
          0x61d000           0x61e000 r--p     1000 1d000  /bin/ls
          0x61e000           0x61f000 rw-p     1000 1e000  /bin/ls
          0x61f000           0x620000 rw-p     1000 0      [heap]
    0x7ffff715a000     0x7ffff7172000 r-xp    18000 0      /lib/x86_64-linux-gnu/libpthread-2.23.so
    0x7ffff7172000     0x7ffff7371000 ---p   1ff000 18000  /lib/x86_64-linux-gnu/libpthread-2.23.so
    0x7ffff7371000     0x7ffff7372000 r--p     1000 17000  /lib/x86_64-linux-gnu/libpthread-2.23.so
    0x7ffff7372000     0x7ffff7373000 rw-p     1000 18000  /lib/x86_64-linux-gnu/libpthread-2.23.so
    0x7ffff7373000     0x7ffff7377000 rw-p     4000 0      
    0x7ffff7377000     0x7ffff737a000 r-xp     3000 0      /lib/x86_64-linux-gnu/libdl-2.23.so
    0x7ffff737a000     0x7ffff7579000 ---p   1ff000 3000   /lib/x86_64-linux-gnu/libdl-2.23.so
    0x7ffff7579000     0x7ffff757a000 r--p     1000 2000   /lib/x86_64-linux-gnu/libdl-2.23.so
    0x7ffff757a000     0x7ffff757b000 rw-p     1000 3000   /lib/x86_64-linux-gnu/libdl-2.23.so
    0x7ffff757b000     0x7ffff75e9000 r-xp    6e000 0      /lib/x86_64-linux-gnu/libpcre.so.3.13.2
    0x7ffff75e9000     0x7ffff77e9000 ---p   200000 6e000  /lib/x86_64-linux-gnu/libpcre.so.3.13.2
    0x7ffff77e9000     0x7ffff77ea000 r--p     1000 6e000  /lib/x86_64-linux-gnu/libpcre.so.3.13.2
    0x7ffff77ea000     0x7ffff77eb000 rw-p     1000 6f000  /lib/x86_64-linux-gnu/libpcre.so.3.13.2
    0x7ffff77eb000     0x7ffff79ab000 r-xp   1c0000 0      /lib/x86_64-linux-gnu/libc-2.23.so
    0x7ffff79ab000     0x7ffff7bab000 ---p   200000 1c0000 /lib/x86_64-linux-gnu/libc-2.23.so
    0x7ffff7bab000     0x7ffff7baf000 r--p     4000 1c0000 /lib/x86_64-linux-gnu/libc-2.23.so
    0x7ffff7baf000     0x7ffff7bb1000 rw-p     2000 1c4000 /lib/x86_64-linux-gnu/libc-2.23.so
    0x7ffff7bb1000     0x7ffff7bb5000 rw-p     4000 0      
    0x7ffff7bb5000     0x7ffff7bd4000 r-xp    1f000 0      /lib/x86_64-linux-gnu/libselinux.so.1
    0x7ffff7bd4000     0x7ffff7dd3000 ---p   1ff000 1f000  /lib/x86_64-linux-gnu/libselinux.so.1
    0x7ffff7dd3000     0x7ffff7dd4000 r--p     1000 1e000  /lib/x86_64-linux-gnu/libselinux.so.1
    0x7ffff7dd4000     0x7ffff7dd5000 rw-p     1000 1f000  /lib/x86_64-linux-gnu/libselinux.so.1
    0x7ffff7dd5000     0x7ffff7dd7000 rw-p     2000 0      
    0x7ffff7dd7000     0x7ffff7dfd000 r-xp    26000 0      /lib/x86_64-linux-gnu/ld-2.23.so
    0x7ffff7fcf000     0x7ffff7fd4000 rw-p     5000 0      
    0x7ffff7ff7000     0x7ffff7ffa000 r--p     3000 0      [vvar]
    0x7ffff7ffa000     0x7ffff7ffc000 r-xp     2000 0      [vdso]
    0x7ffff7ffc000     0x7ffff7ffd000 r--p     1000 25000  /lib/x86_64-linux-gnu/ld-2.23.so
    0x7ffff7ffd000     0x7ffff7ffe000 rw-p     1000 26000  /lib/x86_64-linux-gnu/ld-2.23.so
    0x7ffff7ffe000     0x7ffff7fff000 rw-p     1000 0      
    0x7ffffffde000     0x7ffffffff000 rw-p    21000 0      [stack]
0xffffffffff600000 0xffffffffff601000 r-xp     1000 0      [vsyscall]
```
Main (non-PIE) binary:
```
pwndbg> xinfo 0x4048c5
Extended information for virtual address 0x4048c5:

  Containing mapping:
          0x400000           0x41e000 r-xp    1e000 0      /bin/ls

  Offset information:
         Mapped Area 0x4048c5 = 0x400000 + 0x48c5
         File (Base) 0x4048c5 = 0x400000 + 0x48c5
      File (Segment) 0x4048c5 = 0x400000 + 0x48c5
         File (Disk) 0x4048c5 = /bin/ls  + 0x48c5

 Containing ELF sections:
               .text 0x4048c5 = 0x402a00 + 0x1ec5
pwndbg> xinfo 0x61e012
Extended information for virtual address 0x61e012:

  Containing mapping:
          0x61e000           0x61f000 rw-p     1000 1e000  /bin/ls

  Offset information:
         Mapped Area 0x61e012 = 0x61e000 + 0x12
         File (Base) 0x61e012 = 0x400000 + 0x21e012
      File (Segment) 0x61e012 = 0x61de00 + 0x212
         File (Disk) 0x61e012 = /bin/ls  + 0x1e012

 Containing ELF sections:
            .got.plt 0x61e012 = 0x61e000 + 0x12
pwndbg> xinfo 0x61e610
Extended information for virtual address 0x61e610:

  Containing mapping:
          0x61e000           0x61f000 rw-p     1000 1e000  /bin/ls

  Offset information:
         Mapped Area 0x61e610 = 0x61e000 + 0x610
         File (Base) 0x61e610 = 0x400000 + 0x21e610
      File (Segment) 0x61e610 = 0x61de00 + 0x810
         File (Disk) 0x61e610 = [not file backed]

 Containing ELF sections:
                .bss 0x61e610 = 0x61e600 + 0x10
```
PIE shared library:
```
pwndbg> xinfo 0x7ffff77eb440
Extended information for virtual address 0x7ffff77eb440:

  Containing mapping:
    0x7ffff77eb000     0x7ffff79ab000 r-xp   1c0000 0      /lib/x86_64-linux-gnu/libc-2.23.so

  Offset information:
         Mapped Area 0x7ffff77eb440 = 0x7ffff77eb000 + 0x440
         File (Base) 0x7ffff77eb440 = 0x7ffff77eb000 + 0x440
      File (Segment) 0x7ffff77eb440 = 0x7ffff77eb000 + 0x440
         File (Disk) 0x7ffff77eb440 = /lib/x86_64-linux-gnu/libc-2.23.so + 0x440

 Containing ELF sections:
           .gnu.hash 0x7ffff77eb440 = 0x7ffff77eb2b8 + 0x188
```

Commit:
Fix bug: Reference mem_end instead of file_end

jonaskaempf commented:
Regarding memoization, we should be able to memoize.forever on a given ELF. Also this allows a memoized answer to `pwndbg.elf.get_elf_info("/bin/ls").is_pic`. @disconnect3d I think that addresses your 2 comments above?

Btw the "is pie" information appears to be duplicated in `wrappers.checksec.pie_status()`?

As an aside I noted that `wrappers.readelf.get_jmpslots()` also contains some ELF parsing, but with a subprocess to `readelf`. It may make sense to merge that into `elf.get_elf_info()`. For now, `elf.get_elf_info()` returns pretty basic data structures containing some ELF metadata. The main goal was to extract the needed metadata from the pyelftools parser without needing the open file stream that pyelftools parser requires. In the future, it might be smart to look at some of the things [pwntools](https://github.com/Gallopsled/pwntools/tree/dev/pwnlib/elf) have implemented.

disconnect3d commented:
I don't like the idea of `memoize.forever on a given ELF` as the ELF might change outside of pwndbg. If the operation isn't of super high cost I would stay with invalidating the cache for each reload of that elf.

> The main goal was to extract the needed metadata from the pyelftools parser without needing the open file stream that pyelftools parser requires. In the future, it might be smart to look at some of the things pwntools have implemented.

Yeah, totally agree. The only problem with pwntools is that they don't support Py3 [yet?].

disconnect3d commented on pwndbg/elf.py:
```diff
@@ -68,6 +84,95 @@ def read(typ, address, blob=None):
+    _headers = dict(raw_info.header)
+    _headers['e_entry'] += load
+
+    _segments = []
```
Can we remove underscores before variable names here? In many cases underscore before a local variable is used to indicate the variable is not used at all - e.g. when looping, when you want to discard particular variable and tell IDE to stop informing you the variable is not used.

jonaskaempf replied:
Did not know that. Done

disconnect3d commented on pwndbg/elf.py:
```diff
@@ -68,6 +84,95 @@ def read(typ, address, blob=None):
     return obj
 
 
+@pwndbg.memoize.forever
```
Can we change it to memoize on new objfile?

jonaskaempf replied:
Done

disconnect3d commented on pwndbg/elf.py:
```diff
@@ -68,6 +84,95 @@ def read(typ, address, blob=None):
+        return ELFInfo(header, sections, segments)
+
+
+@pwndbg.memoize.forever
```
Can we change it to memoize on new objfile?

jonaskaempf replied:
Done

disconnect3d commented on pwndbg/commands/xinfo.py:
```diff
@@ -63,17 +63,38 @@ def xinfo_mmap_file(page, addr):
+    for segment in containing_loads:
+        if segment['p_type'] == 'PT_LOAD' and addr < segment['x_vaddr_file_end']:
+            file_offset = segment['p_offset'] + (addr - segment['p_vaddr'])
+            print_line("File (Disk)", addr, file_name, file_offset, "+")
```
There is one whitespace too much on one of my systems, idk why; I am fine to leave it as it is as it is not that important but if u have time to look at it, it would be nice

```
  Offset information:
         Mapped Area 0x4048c5 = 0x400000 + 0x48c5
         File (Base) 0x4048c5 = 0x400000 + 0x48c5
      File (Segment) 0x4048c5 = 0x400000 + 0x48c5
         File (Disk) 0x4048c5 = /bin/ls   + 0x48c5
```

jonaskaempf replied:
Cannot reproduce.

Have tried with ELFs with short and long names, and with small (0x400000) and large (0x555555554000) load addresses. All produce consistent results: Space padding up to the length of the address above, or else just a single space before the "+".

The code in `print_line` for formatting the filename looks sound, and seems to work consistently across py2 and 3 on my system.

Could you try doing `file_name.strip()` on systems that show this bug? Cannot see how it would be the case that some spaces are hiding there without breaking stuff, but currently have no other leads.

Hmm, also not entirely sure about py2 vs 3 'type()' semantics - could this line

```
M.get(first) if type(first) is not str else first.ljust(len(hex(addr)))
```

sometimes result in `M.get()` rendering `first` instead of `first.ljust(len(hex(addr)))` for whatever type `page.objfile` has?

disconnect3d replied:
It is better to do `not isinstance(first, str)`

disconnect3d replied:
The problem with formatting is on Py2 as `hex(addr)` might return an `L` at the end:
```
(Pdb) hex(addr)
'0x4048c5L'
```

disconnect3d replied:
I have fixed both things (`isinstance` and `.rstrip('L')`).

disconnect3d reviewed (changes requested):
see comments

Commit:
Don't use underscore variable names; change decorator to reset_on_objfile

disconnect3d committed:
Update xinfo.py
